core/windows_and_tabs/windows_and_tabs.py

```diff
@@ -0,0 +1,48 @@
+from talon import Context, actions, ui
+
+ctx = Context()
+
+
+@ctx.action_class("app")
+class AppActions:
+    def window_previous():
+        cycle_windows(ui.active_app(), -1)
+
+    def window_next():
+        cycle_windows(ui.active_app(), 1)
+
+
+def cycle_windows(app: ui.App, diff: int):
+    """Cycle windows backwards or forwards for the given application"""
+    active = app.active_window
+    windows = [w for w in app.windows() if w == active or is_window_valid(w)]
+    windows.sort(key=lambda w: w.id)
+    current = windows.index(active)
+    max = len(windows) - 1
+    i = cycle(current + diff, 0, max)
+
+    while i != current:
+        try:
+            actions.user.switcher_focus_window(windows[i])
+            break
+        except Exception:
+            i = cycle(i + diff, 0, max)
+
+
+def is_window_valid(window: ui.Window) -> bool:
+    """Returns true if this window is valid for focusing"""
+    return (
+        not window.hidden
+        and window.title != ""
+        and window.rect.width > window.screen.dpi
+        and window.rect.height > window.screen.dpi
```

Comment:
```suggestion
        # This excludes many tiny windows that are not actual windows, and is a rough heuristic.
        and window.rect.width > window.screen.dpi
        and window.rect.height > window.screen.dpi
```

Reply:
I think this is another case where on macOS, window widths are reported in resolution-independent points, so comparing it against dots per inch doesn't make sense.

Also @nriley -- do you know if there's some property on the window in macOS that we could be checking, to exclude it from this kind of logic? This just made me remember that I already run to this problem in other cases where I use the Talon window snapping code.

For example, the find dialog in chrome counts as a "window":

![image](https://github.com/talonhub/community/assets/536668/a1416923-e9e4-496a-9d3e-2acf80fba713)

... but macOS knows to exclude it from cmd-`. If there's an OS property we could be checking we could exclude it from window snapping logic in other places, subject to the user's preferences, naturally.